### ISSUE TITLE
History: Unique columns

### ISSUE THREAD

**Issue by harveyappleton** [anchan828/typeorm-helpers#656]
Hey @anchan828 

Just ran into an issue, if there any columns marked as `unique: true` in the original entity, it will mean you can only add 1 log file entry before you end up with an SQL error due to a unique constraint.

Is there a way we can add a new decorator that we can apply to the history entities that removes all unique constraints for that entity? Something like `@RemoveAllUnique()` ? Could follow a similar suit to the code for the typeorm `@Unique` column? https://github.com/typeorm/typeorm/blob/master/src/decorator/Unique.ts#L27

Thanks!

**Comment by anchan828** [comment]
I agree. This problem is unavoidable when the schema is automatically generated by `synchronize: true`.

I will add new decorator `@HistoryEntity`. And you will be able to set options like `@HistoryEntity({ removeAllUniques: true})`

```ts
@HistoryEntity({ removeAllUniques: true})
class TestHistoryEntity extends TestEntity {
  @HistoryOriginalIdColumn()
  public originalID!: number;

  @HistoryActionColumn()
  public action!: HistoryActionType;

  @PrimaryGeneratedColumn()
  public id!: number;
}
```

If it doesn't work, come up with another idea 😉

**Comment by harveyappleton** [comment]
Amazing thanks @anchan828 . Sounds like a good solution, let me know if that isn't possible for any reason, but seems ideal!

**Comment by anchan828** [comment]
hmm... In the case of inheritance, it seems that the actual unique object is not defined. 🤔 
This may take some time to resolve.

```ts
export interface HistoryEntityOptions extends EntityOptions {
  removeAllUniques?: boolean;
}

export function HistoryEntity(options?: HistoryEntityOptions): ClassDecorator {
  return (target: Function) => {
    Entity(options)(target);
    console.log(getMetadataArgsStorage().filterUniques(target)); // => []
  };
}
```

**Comment by anchan828** [comment]
I wish there was a hook like preSynchronizeHook that would be called right before synchronization.

**Comment by anchan828** [comment]
You needs to prepare a migration. Therefore, we will provide a helper function so that we can easily remove the unique index. #657

**Comment by anchan828** [comment]
Published https://www.npmjs.com/package/@anchan828/typeorm-history/v/0.5.2 🥳 
If this solves your problem, it would be helpful if you could close the issue
<img width="709" alt="c525c99a1221386a04ea7a915afc5ebf" src="https://user-images.githubusercontent.com/694454/125507133-d9361328-246c-4d3b-9b9b-7b278d3671bc.png">

**Comment by harveyappleton** [comment]
Thanks @anchan828 , will check it out 👍 obviously this only works for migrations and not for when TypeORM `synchronize: true` is on in connection options?

**Comment by anchan828** [comment]
Looks like it's working fine 🤔 
https://github.com/anchan828/typeorm-helpers/blob/8d0d7c87d15c122f220cacb2a5507fab85be3690/packages/history/src/history-entity.spec-e2e.ts#L768

**Comment by harveyappleton** [comment]
@anchan828 just tried it, looks good with the exception of this error:

```
QueryFailedError: Cannot drop index 'REL_81121663f07b1ca6dc7e6ba641': needed in a foreign key constraint
    at new QueryFailedError (/src/error/QueryFailedError.ts:11:9)
    at Query.onResult (/src/driver/mysql/MysqlQueryRunner.ts:193:37)
    at Query.execute (/node_modules/mysql2/lib/commands/command.js:30:14)
    at PoolConnection.handlePacket (/node_modules/mysql2/lib/connection.js:425:32)
    at PacketParser.onPacket (/node_modules/mysql2/lib/connection.js:75:12)
    at PacketParser.executeStart (/node_modules/mysql2/lib/packet_parser.js:75:16)
    at Socket.<anonymous> (/node_modules/mysql2/lib/connection.js:82:25)
    at Socket.emit (node:events:394:28)
    at addChunk (node:internal/streams/readable:312:12)
    at readableAddChunk (node:internal/streams/readable:287:9) {
  code: 'ER_DROP_INDEX_FK',
  errno: 1553,
  sqlState: 'HY000',
  sqlMessage: "Cannot drop index 'REL_81121663f07b1ca6dc7e6ba641': needed in a foreign key constraint",
  query: 'DROP INDEX `REL_81121663f07b1ca6dc7e6ba641` ON `table_name_was_here`',
  parameters: []
}
```

This is a `@OneToOne` relation (tagged with a `@JoinColumn`) that is throwing this error. Any thoughts?

**Comment by harveyappleton** [comment]
(By the way, I was running that migration both **with** and **without** the third argument (`keepIndex: true`), no difference I'm afraid)

**Comment by anchan828** [comment]
I will check tonight!

**Comment by anchan828** [comment]
I think I was able to fix it #658
https://www.npmjs.com/package/@anchan828/typeorm-history/v/0.5.3

It's getting pretty hacky 😂 I think you should write your own migration for further support 🤔

**Comment by harveyappleton** [comment]
Perfect, thanks so much @anchan828 ! You're a hero 🎉
